yt_dlp/utils.py

```diff
@@ -5298,3 +5301,75 @@ def all_args(self):
 
     def parse_args(self):
         return self._parser.parse_args(list(self.all_args))
+
+
+# taken from https://github.com/python/cpython/blob/3.9/Lib/asyncio/runners.py with modifications
+def run_with_loop(main, loop):
+    if asyncio.events._get_running_loop() is not None:
+        raise RuntimeError(
+            "asyncio.run() cannot be called from a running event loop")
+
+    if not asyncio.coroutines.iscoroutine(main):
+        raise ValueError("a coroutine was expected, got {!r}".format(main))
+
+    try:
+        # asyncio.events.set_event_loop(loop)
+        return loop.run_until_complete(main)
+    finally:
+        loop.run_until_complete(loop.shutdown_asyncgens())
+        if hasattr(loop, 'shutdown_default_executor'):
+            loop.run_until_complete(loop.shutdown_default_executor())
```

Comment:
there is already a limited `compat_asyncio_run` function. It can be modified with this full definition

Reply:
In that case, since we don't want any other code to call this, `run_with_loop` can be perhaps be moved inside WebsocketsWrapper